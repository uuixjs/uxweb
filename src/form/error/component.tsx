import { AlignItems, Display, FontSize, Layout } from "../../layout";
import { styled, themeTokenRule } from "@uuixjs/uuixweb-lib";

import { FC } from "react";
import { Icon } from "../../icon";
import { SVGAsset } from "../../svg";
import { withOverlayContext } from "../../overlay-region/context";

export interface FormErrorProps {
  /**
   * @example The value you entered is invalid.
   */
  errorMessage: string;
  overlay?: boolean;
}

const ScFormError = styled(Layout)<{ $overlay?: FormErrorProps["overlay"] }>`
  color: ${({ $overlay }) =>
    $overlay
      ? themeTokenRule("color-text-overlay-error")
      : themeTokenRule("color-text-error")};
`;

export const FormErrorComponent: FC<FormErrorProps> = ({
  errorMessage,
  overlay,
  ...props
}) => {
  return (
    <ScFormError
      display={Display.Flex}
      fontSize={FontSize.Size7}
      $overlay={overlay}
      {...props}
    >
      <Layout
        margin={{ right: 0.5 }}
        display={Display.Flex}
        alignItems={AlignItems.Baseline}
        height="1.5em"
        width="1.5em"
      >
        <Icon fill asset={SVGAsset.NotificationError} />
      </Layout>
      <p>{errorMessage}</p>
    </ScFormError>
  );
};

FormErrorComponent.displayName = "FormErrorComponent";
export const FormError = withOverlayContext(FormErrorComponent);
