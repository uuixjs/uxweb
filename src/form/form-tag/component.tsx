import {
  CoreInteractive,
  CoreInteractivePublicProps,
} from "../../core-interactive";
import { SVG, SVGAsset } from "../../svg";
import {
  css,
  focusVisible,
  getAriaProps,
  getDataProps,
  hoverCss,
  staticTokenRule,
  styleVariant,
  styled,
  themeTokenRule,
} from "@uuixjs/uuixweb-lib";

import { FC } from "react";
import { InputSize } from "../form";
import { getInputSizeValue } from "../common";

export enum FormTagStatus {
  Unselected = "unselected",
  Selected = "selected",
  Locked = "locked",
}

export interface FormTagProps extends CoreInteractivePublicProps {
  /** Text to display within the tag. */
  label: string;
  /** Size of the tag. */
  size?: InputSize;
  status?: FormTagStatus | "unselected" | "selected" | "locked"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
}

interface ScFormTagProps {
  size: InputSize;
  disabled?: boolean;
  status: FormTagStatus | "unselected" | "selected" | "locked"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
}

const ScFormTag = styled.div<ScFormTagProps>`
  /* Input reset */
  font-family: inherit;
  appearance: none;

  display: flex;
  align-items: center;
  border-radius: ${staticTokenRule("border-radius-rounded")};
  background-clip: padding-box;
  color: ${themeTokenRule("color-text-tag")};
  line-height: 1.4;
  white-space: nowrap;
  padding-right: ${({ size }) =>
    size === InputSize.Large ? "0.6rem" : "0.3rem"};
  padding-left: 0.8rem;
  height: ${getInputSizeValue};
  border-width: ${staticTokenRule("border-width-tag")};
  border-style: ${({ status }) =>
    status === FormTagStatus.Unselected ? "dashed" : "solid"};
  border-color: ${themeTokenRule("color-border-tag")};

  background-color: ${styleVariant("status", {
    [FormTagStatus.Unselected]: "transparent",
    [FormTagStatus.Selected]: themeTokenRule("color-background-tag-default"),
    [FormTagStatus.Locked]: themeTokenRule("color-background-tag-default"),
  })};

  /* Disabled styles */
  ${({ disabled }) =>
    disabled &&
    css`
      &,
      &:disabled {
        cursor: not-allowed;
        opacity: 0.5;
      }
    `};
`;

const ScFormTagLocked = styled(ScFormTag)<ScFormTagProps>`
  cursor: default;
`;

const ScFormTagDefault = styled(ScFormTag)<ScFormTagProps>`
  /* Interactive styles */
  ${({ disabled }) =>
    !disabled &&
    css`
      ${hoverCss`
        background: ${themeTokenRule(
          "color-background-interactable-alpha-hover",
        )};
        color: ${themeTokenRule("color-text-tag")};
        text-decoration: none;
      `}

      ${focusVisible`
        background: ${themeTokenRule(
          "color-background-interactable-alpha-hover",
        )};
        border-style: solid;
        border-color: ${themeTokenRule("color-border-input-focus")}
      `}

      &:active {
        background: ${themeTokenRule(
          "color-background-interactable-alpha-active",
        )};
      }
    `};
`;

const ScFormTagLayout = styled.div<Pick<ScFormTagProps, "size">>`
  display: flex;
  align-items: center;
  border-radius: ${staticTokenRule("border-radius-rounded")};
  font-weight: ${staticTokenRule("font-weight-semibold")};
  font-size: ${styleVariant("size", {
    [InputSize.Default]: staticTokenRule("font-size-6"),
    [InputSize.Small]: staticTokenRule("font-size-7"),
    [InputSize.Large]: staticTokenRule("font-size-5"),
  })};
`;

const ScFormTagIcon = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: ${staticTokenRule("border-radius-rounded")};
  width: 1.6rem;
  height: 1.6rem;
`;

export const FormTag: FC<FormTagProps> = ({
  label,
  size = InputSize.Default,
  status = FormTagStatus.Unselected,
  disabled,
  ...formTagProps
}) => {
  const className = "tw-form-tag";
  const FORMTAG_STATUS_ASSET = {
    [FormTagStatus.Unselected]: SVGAsset.Smallplus,
    [FormTagStatus.Selected]: SVGAsset.Smallhide,
    [FormTagStatus.Locked]: SVGAsset.Smalllock,
  };
  const styledProps = {
    size,
    disabled,
    status,
  };

  const asset = FORMTAG_STATUS_ASSET[status];

  const formTagLayout = (
    <ScFormTagLayout size={size}>
      {label}
      <ScFormTagIcon>
        <SVG asset={asset} fill />
      </ScFormTagIcon>
    </ScFormTagLayout>
  );

  if (status === FormTagStatus.Locked) {
    return (
      <ScFormTagLocked
        className={className}
        {...styledProps}
        {...getAriaProps(formTagProps)}
        {...getDataProps(formTagProps)}
      >
        {formTagLayout}
      </ScFormTagLocked>
    );
  } else {
    return (
      <ScFormTagDefault
        className={className}
        as={CoreInteractive}
        {...styledProps}
        {...formTagProps}
        {...getDataProps(formTagProps)}
      >
        {formTagLayout}
      </ScFormTagDefault>
    );
  }
};

FormTag.defaultProps = {
  status: FormTagStatus.Unselected,
  size: InputSize.Default,
};

FormTag.displayName = "FormTag";
