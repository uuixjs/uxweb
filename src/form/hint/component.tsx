import { Color, FontSize } from "../../layout";
import { CoreText, TextType } from "../../core-text";

import { FC } from "react";
import { getAriaProps } from "@uuixjs/uuixweb-lib";
import { withOverlayContext } from "../../overlay-region/context";

export interface FormHintProps {
  /**
   * @example Any combination of letters or numbers
   */
  hint: string;
  overlay?: boolean;
}

export const FormHintComponent: FC<FormHintProps> = (props) => {
  return (
    <CoreText
      type={TextType.P}
      fontSize={FontSize.Size7}
      color={props.overlay ? Color.OverlayAlt : Color.Alt2}
      {...getAriaProps(props)}
    >
      {props.hint}
    </CoreText>
  );
};

FormHintComponent.displayName = "FormHintComponent";

export const FormHint = withOverlayContext(FormHintComponent);
