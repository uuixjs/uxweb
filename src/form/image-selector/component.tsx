import {
  AlignItems,
  Color,
  Display,
  JustifyContent,
  Layout,
  LayoutProps,
  Position,
} from "../../layout";
import { Aspect, AspectRatio } from "../../aspect";
import {
  BorderRadius,
  a11yHide,
  css,
  getAriaProps,
  getBorderRadiusStyles,
  getDataProps,
  hoverCss,
  hoverCssWithSelector,
  newUUIDv4,
  staticTokenRule,
  styleVariant,
  styled,
  themeTokenRule,
} from "lib";
import { CoreImage, CoreImageProps } from "../../core-image";

import { Component } from "react";
import { Icon } from "../../icon";
import { InputControlProps } from "../form";
import { SVGAsset } from "../../svg";
import { getImageProps } from "../../../_utils/image-props";
import { withOverlayContext } from "../../overlay-region/context";

export enum ImageSelectorType {
  Checkbox = "checkbox",
  Radio = "radio",
}

export enum ImageSelectorIconPosition {
  TopLeft = 1,
  TopRight = 2,
  BottomLeft = 3,
  BottomRight = 4,
  Center = 5,
}

export interface ImageSelectorIcon {
  position?: ImageSelectorIconPosition;
  icon?: SVGAsset;
}

export interface ImageSelectorProps extends CoreImageProps, InputControlProps {
  /** Sets the image border radius. */
  borderRadius?: BorderRadius;
  /** ImageSelectors can behave a radio or checkbox controls. */
  type: ImageSelectorType | "checkbox" | "radio"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
  /** Animates image scale on hover. */
  hoverScale?: boolean;
  /** Displays a semi-opaque mask on selected images */
  selectedMask?: boolean;
  /** Displays an icon on selected images. Both icon and position are configurable. Default icon is a circle-check and position is TopRight. */
  selectedIcon?: boolean | ImageSelectorIcon;
  /** Sets the aspect ratio of the image in the selector. */
  ratio?: AspectRatio;
  /** Sets overlay representation */
  overlay?: boolean;
}

interface ScImageSelectorProps {
  fullWidth?: boolean;
}

const ScImageSelector = styled.div<ScImageSelectorProps>`
  width: ${({ fullWidth }) => fullWidth && "100%"};
  position: relative; /* wraps a position: absolute later on */
`;

const ScImageSelectorInput = styled.input`
  ${a11yHide}
`;

interface ScImageSelectorLabelProps {
  borderRadius?: BorderRadius;
  error?: boolean;
  overlay?: boolean;
}

const ScImageSelectorLabel = styled.label<ScImageSelectorLabelProps>`
  display: block;
  position: relative;
  padding: 0.2rem;
  cursor: pointer;
  ${getBorderRadiusStyles}
  border: ${staticTokenRule("border-width-default")} solid
    ${({ error, overlay }) =>
      error
        ? themeTokenRule("color-border-error")
        : overlay
        ? themeTokenRule("color-border-image-selector-overlay")
        : themeTokenRule("color-border-image-selector")};
  box-shadow: ${({ error }) =>
    error
      ? css`inset 0 0 0 0.1rem ${themeTokenRule("color-border-error")}`
      : undefined};

  /* Hover styles */
  ${hoverCss`
    border: ${staticTokenRule("border-width-default")} solid
      ${({ error, overlay }) =>
        error
          ? themeTokenRule("color-border-error")
          : overlay
          ? themeTokenRule("color-border-image-selector-overlay-selected")
          : themeTokenRule("color-border-input-checkbox-checked")};
  `}

  /* Focus styles */
  ${ScImageSelectorInput}[data-focus-visible-added] + && {
    border: ${staticTokenRule("border-width-default")} solid
      ${({ error, overlay }) =>
        error
          ? themeTokenRule("color-border-error")
          : overlay
          ? themeTokenRule("color-border-image-selector-overlay-selected")
          : themeTokenRule("color-border-input-checkbox-checked")};
    box-shadow: 0 0 ${staticTokenRule("border-radius-large")}
      ${({ error, overlay }) =>
        error
          ? themeTokenRule("color-border-error")
          : overlay
          ? themeTokenRule("color-border-image-selector-overlay-selected")
          : themeTokenRule("color-border-input-checkbox-checked")};
  }

  /* Disabled styles */
  ${ScImageSelectorInput}:disabled + && {
    opacity: 0.5;
    pointer-events: none;
  }

  /* Checked styles */
  ${ScImageSelectorInput}:checked + && {
    border: ${staticTokenRule("border-width-default")} solid
      ${({ overlay }) =>
        overlay
          ? themeTokenRule("color-border-image-selector-overlay-selected")
          : themeTokenRule("color-border-input-checkbox-checked")};
    background-color: ${({ overlay }) =>
      overlay
        ? themeTokenRule("color-background-image-selector-overlay")
        : themeTokenRule("color-background-base")};
    box-shadow: inset 0 0 0 0.1rem ${({ overlay }) =>
      overlay
        ? themeTokenRule("color-border-image-selector-overlay-selected")
        : themeTokenRule("color-border-input-checkbox-checked")};
  }

  ${ScImageSelectorInput}[data-focus-visible-added]:checked + && {
    box-shadow: ${({ overlay }) =>
      css`
        0 0 ${staticTokenRule("border-radius-large")} ${
        overlay
          ? themeTokenRule("color-border-image-selector-overlay-selected")
          : themeTokenRule("color-border-input-checkbox-checked")
      },
        inset 0 0 0 0.1rem ${
          overlay
            ? themeTokenRule("color-border-image-selector-overlay-selected")
            : themeTokenRule("color-border-input-checkbox-checked")
        }
      `};
  }
`;

const ScImageSelectorMask = styled.div`
  display: none;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: ${themeTokenRule("color-background-accent")};
  opacity: 0.6;

  ${ScImageSelectorInput}:checked + ${ScImageSelectorLabel} & {
    display: block;
  }
`;

const ScImageSelectorImageBorder = styled.div<ScImageSelectorLabelProps>`
  position: relative;
  overflow: hidden;

  border-radius: ${styleVariant("borderRadius", {
    [BorderRadius.None]: staticTokenRule("border-radius-none"),
    [BorderRadius.Small]: staticTokenRule("border-radius-none"),
    [BorderRadius.Medium]: css`calc(${staticTokenRule(
      "border-radius-medium",
    )} / 2)`,
    [BorderRadius.Large]: css`calc(${staticTokenRule(
      "border-radius-large",
    )} / 2)`,
    [BorderRadius.Rounded]: staticTokenRule("border-radius-rounded"),
  })};
`;

const ScImageSelectorImage = styled(CoreImage)<{ $hoverScale?: boolean }>`
  transition: transform 150ms ease-out;

  ${({ $hoverScale }) =>
    $hoverScale &&
    hoverCssWithSelector(`${ScImageSelector}:hover &`)`
      transform: scale(1.1);
  `};
`;

export class ImageSelectorComponent extends Component<ImageSelectorProps, {}> {
  public static defaultProps: Partial<ImageSelectorProps> = {
    borderRadius: BorderRadius.Large,
  };
  public static displayName: string;

  public render() {
    const generatedId = newUUIDv4();

    return (
      <ScImageSelector
        className="tw-image-selector"
        fullWidth={this.props.ratio !== undefined}
        {...getDataProps(this.props)}
      >
        <ScImageSelectorInput
          {...getAriaProps(this.props)}
          className="tw-image-selector__input"
          type={this.props.type}
          data-a-target="tw-image-selector"
          checked={this.props.checked}
          defaultChecked={this.props.defaultChecked}
          disabled={this.props.disabled}
          id={this.props.id ? this.props.id : generatedId}
          name={this.props.name}
          onBlur={this.props.onBlur}
          onChange={this.props.onChange}
          onFocus={this.props.onFocus}
          ref={this.props.refDelegate}
          required={this.props.required}
          tabIndex={this.props.tabIndex}
          value={this.props.value}
        />
        <ScImageSelectorLabel
          htmlFor={this.props.id ? this.props.id : generatedId}
          borderRadius={this.props.borderRadius}
          error={this.props.error}
          overlay={this.props.overlay}
        >
          <ScImageSelectorImageBorder borderRadius={this.props.borderRadius}>
            {this.renderImage()}
            {(!!this.props.selectedIcon || !!this.props.selectedMask) &&
              this.renderMask()}
          </ScImageSelectorImageBorder>
        </ScImageSelectorLabel>
      </ScImageSelector>
    );
  }

  private renderImage = () => {
    const image = (
      <ScImageSelectorImage
        className="tw-image-selector__image"
        $hoverScale={this.props.hoverScale}
        {...getImageProps(this.props)}
      />
    );

    if (this.props.ratio === undefined) {
      return image;
    }

    return (
      <Aspect ratio={this.props.ratio || AspectRatio.Aspect1x1}>{image}</Aspect>
    );
  };

  private renderMask = () => {
    return (
      <>
        <ScImageSelectorMask />
        {this.props.selectedIcon && this.renderIcon()}
      </>
    );
  };

  private renderIcon = () => {
    const iconLayoutProps: LayoutProps = {
      color: Color.Overlay,
      position: Position.Absolute,
      margin: 0.5,
    };

    const icon = (
      <Icon
        asset={
          typeof this.props.selectedIcon === "object" &&
          this.props.selectedIcon.icon
            ? this.props.selectedIcon.icon
            : SVGAsset.NotificationSuccess
        }
      />
    );

    if (
      typeof this.props.selectedIcon === "object" &&
      this.props.selectedIcon.position
    ) {
      return (
        <Layout
          {...iconLayoutProps}
          attachTop={
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.Center ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.TopLeft ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.TopRight
          }
          attachRight={
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.Center ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.TopRight ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.BottomRight
          }
          attachBottom={
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.Center ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.BottomLeft ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.BottomRight
          }
          attachLeft={
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.Center ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.TopLeft ||
            this.props.selectedIcon.position ===
              ImageSelectorIconPosition.BottomLeft
          }
          display={
            this.props.selectedIcon.position ===
            ImageSelectorIconPosition.Center
              ? Display.Flex
              : undefined
          }
          alignItems={
            this.props.selectedIcon.position ===
            ImageSelectorIconPosition.Center
              ? AlignItems.Center
              : undefined
          }
          justifyContent={
            this.props.selectedIcon.position ===
            ImageSelectorIconPosition.Center
              ? JustifyContent.Center
              : undefined
          }
        >
          {icon}
        </Layout>
      );
    } else {
      return (
        <Layout
          {...iconLayoutProps}
          display={Display.Flex}
          attachTop
          attachRight
        >
          {icon}
        </Layout>
      );
    }
  };
}

ImageSelectorComponent.displayName = "ImageSelectorComponent";

export const ImageSelector = withOverlayContext(ImageSelectorComponent);
