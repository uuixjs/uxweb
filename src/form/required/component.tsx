import { SVGAsset, SVG_PATHS } from "../../svg";

import { FC } from "react";
import { styled } from "@uuixjs/uuixweb-lib";

const ScFormRequired = styled.svg`
  fill: currentColor;
`;
export const FormRequired: FC<{}> = () => {
  return (
    <ScFormRequired
      className="tw-form-required"
      role="img"
      width="1em"
      height="1em"
      version="1.1"
      viewBox="0 0 12 12"
      x="0"
      y="0"
    >
      {SVG_PATHS[SVGAsset.Xsmallasterisk].path}
    </ScFormRequired>
  );
};

FormRequired.displayName = "FormRequired";
