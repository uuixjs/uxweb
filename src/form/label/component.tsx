import { staticTokenRule, styled, themeTokenRule } from "@uuixjs/uuixweb-lib";

import { FC } from "react";
import { FormRequired } from "../required";
import { withOverlayContext } from "../../overlay-region/context";

export interface FormLabelProps {
  id: string;
  label: string;
  required?: boolean;
  overlay?: boolean;
}

export const ScFormLabel = styled.label<{
  $overlay?: FormLabelProps["overlay"];
}>`
  color: ${({ $overlay }) =>
    $overlay
      ? themeTokenRule("color-text-overlay")
      : themeTokenRule("color-text-label")};
  font-weight: ${staticTokenRule("font-weight-bold")};
`;

export const FormLabelComponent: FC<FormLabelProps> = ({
  id,
  label,
  required,
  overlay,
  ...props
}) => {
  return (
    <ScFormLabel
      className="tw-form-label"
      htmlFor={id}
      $overlay={overlay}
      {...props}
    >
      {required && <FormRequired />}
      {label}
    </ScFormLabel>
  );
};

FormLabelComponent.displayName = "FormLabelComponent";

export const FormLabel = withOverlayContext(FormLabelComponent);
