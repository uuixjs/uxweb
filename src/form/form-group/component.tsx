import {
  Display,
  FlexWrap,
  FontSize,
  Layout,
  LayoutProps,
  Position,
} from "../../layout";
import { FC, ReactNode } from "react";
import {
  newUUIDv4,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "@uuixjs/uuixweb-lib";

import { FormError } from "../error";
import { FormHint } from "../hint";
import { FormLabel } from "../label";
import { FormLegend } from "../legend";

export enum FormGroupOrientation {
  Vertical = "vertical",
  Horizontal = "horizontal",
}

export enum FormGroupControls {
  Single = "single",
  Multiple = "multiple",
}

export interface FormGroupProps {
  controls?: FormGroupControls | "single" | "multiple"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
  /**
   * @example <Input type={InputType.Text} placeholder="Video Title" />
   */
  children?: ReactNode;
  error?: ReactNode;
  /** @deprecated Set an error message on `error` to toggle and display an error message. */
  errorMessage?: string;
  hint?: ReactNode;
  id?: string;
  /**
   * @example Title
   */
  label: ReactNode;
  /** @deprecated Please include this as part of the `label` prop.  */
  labelOptional?: string;
  required?: boolean;
  orientation?: FormGroupOrientation | "vertical" | "horizontal"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
}

export const formGroupHorizontalLabelWidth = "18rem";
export const FORM_GROUP_HORIZONTAL_LABEL_TEST_SELECTOR =
  "tw-form-group__label-container";

export const ScFormGroupOptional = styled.span`
  color: ${themeTokenRule("color-text-label-optional")};
  padding-left: ${staticTokenRule("space-05")};
  font-weight: ${staticTokenRule("font-weight-normal")};
`;

export const FormGroup: FC<FormGroupProps> = ({
  children,
  controls,
  error,
  errorMessage,
  hint,
  id,
  label,
  labelOptional,
  required,
  orientation,
  ...props
}) => {
  let labelElement: ReactNode;
  let optionalLabel: ReactNode;
  let formHint: ReactNode;
  let container: "fieldset" | "div";

  const generatedId = newUUIDv4();

  if (labelOptional) {
    optionalLabel = (
      <ScFormGroupOptional>({labelOptional})</ScFormGroupOptional>
    );
  }

  if (hint && !error) {
    formHint = typeof hint === "string" ? <FormHint hint={hint} /> : hint;
  } else if (error) {
    formHint =
      typeof error === "string" ? (
        <FormError errorMessage={error} />
      ) : error === true && errorMessage ? (
        <FormError errorMessage={errorMessage} />
      ) : (
        error
      );
  }

  if (formHint) {
    formHint = <Layout margin={{ top: 0.5 }}>{formHint}</Layout>;
  }

  if (controls === FormGroupControls.Multiple) {
    labelElement =
      typeof label === "string" ? <FormLegend label={label} /> : label;
    container = "fieldset";
  } else {
    labelElement =
      typeof label === "string" ? (
        <FormLabel
          id={id ? id : generatedId}
          required={required}
          label={label}
        />
      ) : (
        label
      );
    container = "div";
  }

  labelElement = labelElement && (
    <Layout margin={{ bottom: 0.5 }}>
      {labelElement}
      {typeof label === "string" && optionalLabel}
    </Layout>
  );

  const controlsProps: LayoutProps = {
    className: "tw-form-group",
    position: Position.Relative,
    flexGrow: 1,
    fontSize: FontSize.Size6,
    ...props,
  };

  if (orientation === FormGroupOrientation.Horizontal) {
    return (
      <Layout
        as={container}
        display={Display.Flex}
        flexWrap={FlexWrap.NoWrap}
        {...controlsProps}
      >
        <Layout
          data-test-selector={FORM_GROUP_HORIZONTAL_LABEL_TEST_SELECTOR}
          width={formGroupHorizontalLabelWidth}
          flexShrink={0}
          padding={{ right: 2 }}
        >
          {labelElement}
        </Layout>
        <Layout flexGrow={1}>
          {children}
          {formHint}
        </Layout>
      </Layout>
    );
  } else {
    return (
      <Layout as={container} {...controlsProps}>
        {labelElement}
        {children}
        {formHint}
      </Layout>
    );
  }
};

FormGroup.displayName = "FormGroup";
