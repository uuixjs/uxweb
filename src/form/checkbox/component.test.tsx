import { mount, shallow } from "enzyme";
import { setupShallowTest } from "../../../tests/helpers";
import { InputSize } from "../form";
import { CheckBox, CheckBoxComponent, CheckBoxProps } from "./component";

function getRequiredProps(): CheckBoxProps {
  return {
    label: "Test Label",
    id: "test-id",
  };
}

function getOptionalProps(): CheckBoxProps {
  return {
    autoFocus: true,
    checked: true,
    defaultChecked: true,
    disabled: true,
    error: true,
    id: "test-id",
    label: "Test Label",
    name: "Test name",
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onFocus: jest.fn(),
    overlay: true,
    required: true,
    tabIndex: 0,
    value: "Test value",
  };
}

const setupShallow = setupShallowTest(CheckBoxComponent, getRequiredProps);

describe("CheckBox", () => {
  it("renders checkbox component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders checkbox component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a checkbox component with size prop", () => {
    const { wrapper } = setupShallow({ size: InputSize.Large });
    expect(wrapper).toMatchSnapshot();
  });

  it("renders one data-test-selector", () => {
    const SELECTOR = { "data-test-selector": "my-test-string" };
    const wrapper = shallow(
      <CheckBoxComponent {...getRequiredProps()} {...SELECTOR} />,
    );
    expect(wrapper.find(SELECTOR).length).toEqual(1);
  });

  it("renders with indeterminate attr when prop is true", () => {
    const wrapper = mount(<CheckBox {...getRequiredProps()} indeterminate />);
    expect(wrapper.find("input").getDOMNode()).toHaveProperty("indeterminate");
  });

  it("renders correctly when indeterminate prop is updated", () => {
    const wrapper = mount(<CheckBox {...getRequiredProps()} indeterminate />);
    const checkboxElement = wrapper.find("input").getDOMNode();
    expect(checkboxElement).toHaveProperty("indeterminate", true);
    wrapper.setProps({ indeterminate: false });
    expect(checkboxElement).toHaveProperty("indeterminate", false);
  });

  it("respects refDelegate when indeterminate is true", () => {
    let inputRef = null;
    mount(
      <CheckBox
        {...getRequiredProps()}
        refDelegate={(ref) => (inputRef = ref)}
        indeterminate
      />,
    );
    expect(inputRef).not.toBeNull();
    expect(inputRef).toHaveProperty("type", "checkbox");
  });

  /**
   * It is important to ensure "position: relative" is wrapped
   * around any content which has "position: absolute" applied to it.
   * Without this, there is a bug when the content is rendered inside of
   * Twilight's `<ScrollableArea>` component which causes the scroll
   * region to jump up on the page when clicking on a UI element like
   * a checkbox if there is an un-wrapped position: absolute element involved.
   */
  it("uses position relative on the wrapper and position absolute on the input", () => {
    const wrapper = mount(<CheckBox {...getRequiredProps()} />);

    expect(wrapper).toHaveStyleRule(
      "position",
      expect.stringContaining("relative"),
    );
    expect(wrapper.find("input")).toHaveStyleRule(
      "position",
      expect.stringContaining("absolute"),
    );
  });
});
