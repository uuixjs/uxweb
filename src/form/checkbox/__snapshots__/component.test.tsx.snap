// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox renders a checkbox component with size prop 1`] = `
<Layout
  className="tw-checkbox"
  fontSize="font-size-5"
  position="relative"
>
  <styled.input
    className="tw-checkbox__input"
    data-a-target="tw-checkbox"
    id="test-id"
    type="checkbox"
  />
  <Styled(styled.label)
    className="tw-checkbox__label"
    htmlFor="test-id"
  >
    <Layout
      margin={
        Object {
          "left": 1,
        }
      }
    >
      Test Label
    </Layout>
  </Styled(styled.label)>
</Layout>
`;

exports[`CheckBox renders checkbox component with all props 1`] = `
<Layout
  className="tw-checkbox tw-checkbox--error tw-checkbox--overlay"
  fontSize="font-size-6"
  position="relative"
>
  <styled.input
    autoFocus={true}
    checked={true}
    className="tw-checkbox__input"
    data-a-target="tw-checkbox"
    defaultChecked={true}
    disabled={true}
    id="test-id"
    name="Test name"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    required={true}
    tabIndex={0}
    type="checkbox"
    value="Test value"
  />
  <Styled(styled.label)
    className="tw-checkbox__label"
    error={true}
    htmlFor="test-id"
    overlay={true}
  >
    <Layout
      margin={
        Object {
          "left": 1,
        }
      }
    >
      Test Label
    </Layout>
  </Styled(styled.label)>
</Layout>
`;

exports[`CheckBox renders checkbox component with required props 1`] = `
<Layout
  className="tw-checkbox"
  fontSize="font-size-6"
  position="relative"
>
  <styled.input
    className="tw-checkbox__input"
    data-a-target="tw-checkbox"
    id="test-id"
    type="checkbox"
  />
  <Styled(styled.label)
    className="tw-checkbox__label"
    htmlFor="test-id"
  >
    <Layout
      margin={
        Object {
          "left": 1,
        }
      }
    >
      Test Label
    </Layout>
  </Styled(styled.label)>
</Layout>
`;
